src/mpyl/stages/discovery.py

```diff
@@ -27,89 +40,100 @@ def is_invalidated(
     touched_dependency = (
         next(filter(path.startswith, deps_for_stage), None) if deps else None
     )
-    startswith: bool = path.startswith(project.root_path)
     if touched_dependency:
         logger.debug(
             f"Project {project.name}: {path} touched dependency {touched_dependency}"
         )
-    if startswith:
-        logger.debug(
-            f"Project {project.name}: {path} touched project root {project.root_path}"
-        )
-    return startswith or touched_dependency is not None
-
-
-def output_invalidated(output: Optional[Output], revision_hash: str) -> bool:
-    if output is None:
-        return True
-    if not output.success:
-        return True
-    if output.produced_artifact is None:
-        return True
-    artifact = output.produced_artifact
-    if artifact.revision != revision_hash:
-        return True
-
-    return False
-
-
-def _to_relevant_changes(
-    project: Project, stage: str, change_history: list[Revision]
-) -> set[str]:
-    output: Output = Output.try_read(project.target_path, stage)
-    relevant = set()
-    for history in reversed(sorted(change_history, key=lambda c: c.ord)):
-        if stage == deploy.STAGE_NAME or output_invalidated(output, history.hash):
-            relevant.update(history.files_touched)
-        else:
-            return relevant
+    return touched_dependency is not None
 
-    return relevant
 
+def is_output_cached(output: Optional[Output], cache_key: str) -> bool:
+    if output is None or not output.success or output.produced_artifact is None:
```

Comment:
Like we discussed earlier today: adding an extra condition here like `or not output.produced_artifact.hash` might be good to avoid having to generate a random key in case there are no changes.